### PR TITLE
Fix restart ballistic animation while child size change

### LINKED ISSUE #11599: Ballistic simulations may be created too frequently
Make a normal fling gesture. 
~50% of the time, many goBallistic calls are triggered when I would have expected just one. 

In this example, it took 78ms cumulative recreating the simulations and causes frame drops. 
cw5hukq7xsp

Call stack
```
#0      ScrollPositionWithSingleContext.goBallistic (package:flutter/src/widgets/scroll_position_with_single_context.dart:150:16)
#1      BallisticScrollActivity.applyNewDimensions (package:flutter/src/widgets/scroll_activity.dart:399:14)
#2      ScrollPosition.applyNewDimensions (package:flutter/src/widgets/scroll_position.dart:441:14)
#3      ScrollPositionWithSingleContext.applyNewDimensions (package:flutter/src/widgets/scroll_position_with_single_context.dart:110:29)
#4      ScrollPosition.applyContentDimensions (package:flutter/src/widgets/scroll_position.dart:411:7)
#5      RenderViewport.performLayout (package:flutter/src/rendering/viewport.dart:827:20)
#6      RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:1767:7)
#7      PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:1056:18)
#8      BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:251:19)
#9      BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:504:22)
#10     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:189:5)
#11     BindingBase&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:695:15)
#12     BindingBase&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:643:9)
#13     _invoke (file:///Users/xster/development/flutter/bin/cache/pkg/sky_engine/lib/ui/hooks.dart:86)
#14     _drawFrame (file:///Users/xster/development/flutter/bin/cache/pkg/sky_engine/lib/ui/hooks.dart:75)
```

### PR DESCRIPTION
**Pull request by MxSoul**
Hi, @Piinks
If child size change, performLayout will be call. And goBallistic will be called during performLayout. It will restart ballistic animation. Restarting the animation will cause the ballistic to be inconsistent with the expected curve. 

I don't think the animation should be restarted during a ballistic even if its boundaries change. Maybe we should update related variables instead.

【I already do some test. I will post these data later】

Fix https://github.com/flutter/flutter/issues/11599 & https://github.com/flutter/flutter/issues/24715

This issue https://github.com/flutter/flutter/issues/83632  is also caused by it.
So maybe I will reland the https://github.com/flutter/flutter/pull/77497 after this PR merge.

cc @knopp @nt4f04uNd 
The solution is referenced knopp's comment.  https://github.com/flutter/flutter/issues/24715#issuecomment-683250274

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by Piinks — COMMENTED** [review]
This looks like a great PR, thank you for taking a crack at solving these issues! I should be able to give it a thorough review next week, in the meantime, can you take a look at the failing tests? Is this intended to be a breaking change?

**Comment by MxSoul** [comment]
> This looks like a great PR, thank you for taking a crack at solving these issues! I should be able to give it a thorough review next week, in the meantime, can you take a look at the failing tests? Is this intended to be a breaking change?

I am tring to fix the test case.
It will not be a breaking change. Because I will do something to make it compatible with other simulation.
I am not sure whether the current way to fix is the best. If we do some breaking, maybe we could have a better plan

File: packages/flutter/lib/src/widgets/scroll_simulation.dart [hunk 1]
```diff
@@ -36,13 +45,13 @@ class BouncingScrollSimulation extends Simulation {
...
-       assert(leadingExtent <= trailingExtent),
-       assert(spring != null),
-       super(tolerance: tolerance) {
+  })  : assert(position != null),
+        assert(velocity != null),
+        assert(leadingExtent != null),
+        assert(trailingExtent != null),
+        assert(leadingExtent <= trailingExtent),
+        assert(spring != null),
+        super(tolerance: tolerance) {
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_simulation.dart:54]: The indentation here is intentional in keeping with the style guide, please revert.

File: packages/flutter/lib/src/widgets/scroll_simulation.dart [hunk 2]
```diff
@@ -53,19 +62,21 @@ class BouncingScrollSimulation extends Simulation {
...
+      final double finalX = _frictionSimulation!.finalX;
       if (velocity > 0.0 && finalX > trailingExtent) {
-        _springTime = _frictionSimulation.timeAtX(trailingExtent);
+        _springTime = _frictionSimulation!.timeAtX(trailingExtent);
         _springSimulation = _overscrollSimulation(
           trailingExtent,
-          math.min(_frictionSimulation.dx(_springTime), maxSpringTransferVelocity),
+          math.min(
+              _frictionSimulation!.dx(_springTime), maxSpringTransferVelocity),
         );
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_simulation.dart:72]: Same here and elsewhere, please see the style guide for formatting guidance. The flutter repo does not use dart format.

```suggestion
          math.min(
            _frictionSimulation!.dx(_springTime),
            maxSpringTransferVelocity),
        );
```

File: packages/flutter/lib/src/widgets/scroll_simulation.dart [hunk 3]
```diff
@@ -104,15 +148,33 @@ class BouncingScrollSimulation extends Simulation {
   }
 
   Simulation _simulation(double time) {
-    final Simulation simulation;
+    final Simulation? simulation;
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_simulation.dart:151]: Why does this become nullable?

**Review by Piinks — CHANGES_REQUESTED** [review]
Thanks for working on this @MxSoul, it is not a trivial thing to do as a first-time contributor! (Welcome too! :tada:)

As it currently stands, I am not sure if this implementation will work. Although, I may not be following it fully. Have you considered writing a design document? We usually write up large scale changes in a document so that we can review and give feedback on the concept. It also helps in discussing potential breaks, migration and testing plans, the template is available here: https://docs.google.com/document/d/1SFRO8U2toOlAaZ38dsuEU7Wm5fn41wvBCWKiwADqfmw/edit?usp=sharing

If this is a breaking change, the process is documented in our wiki: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

**Comment by MxSoul** [comment]
> Thanks for working on this @MxSoul, it is not a trivial thing to do as a first-time contributor! (Welcome too! 🎉)
> 
> As it currently stands, I am not sure if this implementation will work. Although, I may not be following it fully. Have you considered writing a design document? We usually write up large scale changes in a document so that we can review and give feedback on the concept. It also helps in discussing potential breaks, migration and testing plans, the template is available here: https://docs.google.com/document/d/1SFRO8U2toOlAaZ38dsuEU7Wm5fn41wvBCWKiwADqfmw/edit?usp=sharing
> 
> If this is a breaking change, the process is documented in our wiki: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

Thanks for your review.
I optimize some implementation And I am writing the design document right now. Maybe it will finish at tomorrow.

**Comment by MxSoul** [comment]
@Piinks Hi~ I write a doc. Please take a look when you have time.
I am not native speaker, it depend on Google Translate. So maybe have some grammar problem.  :P
https://docs.google.com/document/d/1J39wkBaz1lYn7kmT7wuN8mKAvIad98dSOvxzYrDq28Q/edit?resourcekey=0-hCGe2gHb01bP9CvX31TrRQ#

**Comment by MxSoul** [comment]
Hi @Hixie & @nt4f04uNd ,
Maybe someone could help me to review this PR ? Sorry to bother you.

(*[Piinks](https://github.com/Piinks) is out of office right now.)

**Comment by MxSoul** [comment]
@goderbauer Looking forward to your review

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 4]
```diff
@@ -527,26 +532,34 @@ class BallisticScrollActivity extends ScrollActivity {
...
   void resetActivity() {
     delegate.goBallistic(velocity);
   }
 
   @override
   void applyNewDimensions() {
-    delegate.goBallistic(velocity);
+    final ScrollPosition position = delegate.scrollPosition;
+    final dynamic simulation = _simulation;
+    if (position != null && simulation is ScrollSimulationMixin){
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:558]: nit: space between ) and {

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/scroll_activity.dart:558]: The `is ScrollSimulationMixin` is also iffy as was already called out in the previous review, I think?

**Review by goderbauer — COMMENTED** [review]
@MxSoul looks like some of the feedback and questions that @Piinks had haven't been addressed yet. Could you respond to those before requesting another review?

Also, this PR is missing tests. Please add some.

**Comment by Piinks** [comment]
@MxSoul would you like to continue working on this change?

**Comment by Piinks** [comment]
It looks like some of the comments in previous code reviews are unaddressed.

**Comment by Piinks** [comment]
Also, your design document is not open for commenting. Can you update the permissions to allow feedback?

**Comment by MxSoul** [comment]
> Also, your design document is not open for commenting. Can you update the permissions to allow feedback?

OK~ I update the permissions.

**Comment by MxSoul** [comment]
> It looks like some of the comments in previous code reviews are unaddressed.

This is probably because I fixed some of the issues mentioned in the comments and some of the code was removed

File: packages/flutter/lib/src/widgets/scroll_activity.dart (hunk 4 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/scroll_activity.dart:558]: This has not been addressed yet, and cannot be accepted in its current state.

**Comment by skia-gold** [comment]
Gold has detected about 30 new digest(s) on patchset 14.
View them at https://flutter-gold.skia.org/cl/github/96512

**Comment by skia-gold** [comment]
Gold has detected about 32 new digest(s) on patchset 15.
View them at https://flutter-gold.skia.org/cl/github/96512

**Comment by MxSoul** [comment]
I addressed comments and add test. But it make some test break. I will fix it tomorrow

File: packages/flutter/lib/src/physics/spring_simulation.dart [hunk 5]
```diff
@@ -124,27 +124,6 @@ class SpringSimulation extends Simulation {
   String toString() => '${objectRuntimeType(this, 'SpringSimulation')}(end: ${_endPosition.toStringAsFixed(1)}, $type)';
 }
 
-/// A [SpringSimulation] where the value of [x] is guaranteed to have exactly the
```
Comment (Piinks) [inline packages/flutter/lib/src/physics/spring_simulation.dart:127]: Why is this being moved out of the physics library? This would be breaking for anyone that imports it through this library.

File: packages/flutter/lib/src/widgets/list_wheel_scroll_view.dart [hunk 6]
```diff
@@ -479,7 +480,7 @@ class FixedExtentScrollPhysics extends ScrollPhysics {
   }
 
   @override
-  Simulation? createBallisticSimulation(ScrollMetrics position, double velocity) {
+  ScrollSimulation? createBallisticSimulation(ScrollMetrics position, double velocity) {
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/list_wheel_scroll_view.dart:483]: Changes like this would also might be breaking and unable to be merged without a non-breaking migration strategy. The current test failures should be able to confirm this.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 7]
```diff
@@ -1331,7 +1338,7 @@ class _NestedScrollPosition extends ScrollPosition implements ScrollActivityDele
   // ballistic.
   @override
   void goBallistic(double velocity) {
-    Simulation? simulation;
+    ScrollSimulation? simulation;
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart:1341]: Same for all of these regarding breaking changes.

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 8]
```diff
@@ -29,6 +31,9 @@ abstract class ScrollActivityDelegate {
   /// The direction in which the scroll view scrolls.
   AxisDirection get axisDirection;
 
+  /// The current scrollPosition in which the scroll view scrolls.
+  ScrollPosition get scrollPosition;
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:35]: The scroll activity delegate is implemented by a scroll position, why is this needed?

File: packages/flutter/lib/src/widgets/scroll_physics.dart [hunk 9]
```diff
@@ -337,7 +337,7 @@ class ScrollPhysics {
   /// The given `position` is only valid during this method call. Do not keep a
   /// reference to it to use later, as the values may update, may not update, or
   /// may update to reflect an entirely unrelated scrollable.
-  Simulation? createBallisticSimulation(ScrollMetrics position, double velocity) {
+  ScrollSimulation? createBallisticSimulation(ScrollMetrics position, double velocity) {
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_physics.dart:340]: Likewise for all of these return values.

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 10]
```diff
@@ -525,28 +530,35 @@ class BallisticScrollActivity extends ScrollActivity {
...
+
   @override
   void resetActivity() {
     delegate.goBallistic(velocity);
   }
 
   @override
   void applyNewDimensions() {
-    delegate.goBallistic(velocity);
+    final ScrollPosition position = delegate.scrollPosition;
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:556]: Could this change just be implemented in the scroll position, so you do not need to add a position getter to the delegate?

File: packages/flutter/lib/src/widgets/scroll_simulation.dart [hunk 11]
```diff
@@ -6,13 +6,64 @@ import 'dart:math' as math;
 
 import 'package:flutter/foundation.dart';
 import 'package:flutter/physics.dart';
+import 'scroll_activity.dart';
+import 'scroll_position.dart';
+
+/// Extension of ScrollSimulation
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_simulation.dart:12]: When you revisit, can you add more documentation? I am not sure what this means.

**Review by Piinks — COMMENTED** [review]
Thanks for the update, here are just some thoughts for when you return to this. There are still a lot of test failures and analysis issues to resolve before we can review further.

**Comment by MxSoul** [comment]
Hi @Piinks, We may all have some questions to ask each other, because of the time zone, the use of issue communication is a bit inefficient. Could you create a subsection in discord and we use it to communicate.
I will be online at GMT-8 2022.03.09 16:30 (GMT+8 2022.03.10 08:30).

**Comment by Piinks** [comment]
Hey @MxSoul please do feel free to share your questions. I can answer them asynchronously if we are not online at the same time. The #hackers-framework channel is probably a good place to chat, you can start a thread there if you like. :)

**Comment by goderbauer** [comment]
(Triage) closing this one in favor of https://github.com/flutter/flutter/pull/100133.
